Source/AwsCommonRuntimeKit/mqtt/Mqtt5.swift

```diff
@@ -0,0 +1,1053 @@
+            self.serverKeepAliveSec = serverKeepAliveSec
+            self.retainAvailable = retainAvailable
+            self.wildcardSubscriptionsAvailable = wildcardSubscriptionsAvailable
+            self.subscriptionIdentifiersAvailable = subscriptionIdentifiersAvailable
+            self.sharedSubscriptionsAvailable = sharedSubscriptionsAvailable
+            self.rejoinedSession = rejoinedSession
+            self.clientId = clientId
+        }
+}
+
+/// Mqtt5 User Property
+public class UserProperty {
+
+    /// Property name
+    let name: String
+
+    /// Property value
+    let value: String
+
+    init (name: String, value: String) {
+        self.name = name
+        self.value = value
+    }
+}
+
+/// Data model of an `MQTT5 PUBLISH <https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901100>`_ packet
+public class PublishPacket {
+
+    /// The payload of the publish message in a byte buffer format
+    var payload: Data?
+
+    /// The MQTT quality of service associated with this PUBLISH packet.
+    let qos: QoS
+
+    /// The topic associated with this PUBLISH packet.
+    let topic: String
+
+    /// True if this is a retained message, false otherwise.
+    var retain: Bool = false
+
+    /// Property specifying the format of the payload data. The mqtt5 client does not enforce or use this value in a meaningful way.
+    var payloadFormatIndicator: PayloadFormatIndicator?
+
+    /// Sent publishes - indicates the maximum amount of time allowed to elapse for message delivery before the server should instead delete the message (relative to a recipient). Received publishes - indicates the remaining amount of time (from the server's perspective) before the message would have been deleted relative to the subscribing client. If left None, indicates no expiration timeout.
+    var messageExpiryIntervalSec: UInt32?
+
+    /// An integer value that is used to identify the Topic instead of using the Topic Name.  On outbound publishes, this will only be used if the outbound topic aliasing behavior has been set to Manual.
+    var topicAlias: UInt16?
+
+    /// Opaque topic string intended to assist with request/response implementations.  Not internally meaningful to MQTT5 or this client.
+    var responseTopic: String?
+
+    /// Opaque binary data used to correlate between publish messages, as a potential method for request-response implementation.  Not internally meaningful to MQTT5.
+    var correlationData: String? // Unicode objects are converted to C Strings using 'utf-8' encoding
+
+    /// The subscription identifiers of all the subscriptions this message matched.
+    var subscriptionIdentifiers: [UInt32]? // ignore attempts to set but provide in received packets
+
+    /// Property specifying the content type of the payload.  Not internally meaningful to MQTT5.
+    var contentType: String?
+
+    /// Array of MQTT5 user properties included with the packet.
+    var userProperties: [UserProperty]?
```

Comment:
It would be pretty simple to add an appendUserProperty() function but it feels unecessary if `userProperties` is exposed and a customer can simply use `packet.userProperties.append(newUserProperty)`. We could add the function to every packet that contains userProperties but it just felt like a lot of useless code. Definitely different if we don't expose userProperties though. Will put it up as a discussion point.